USB: Add DJ Hero Turntable Emulation

sanjay900 opened the pull request:
### Description of Changes
Add support for emulating DJ hero turntables. Note that this doesn't actually support PS3 turntables, as they require raw HID access, and i wouldn't really know the best approach to that. However, a 360 turntable should work, as should using something like https://github.com/ghlre/GHLtarUtility to emulate a 360 turntable.

### Rationale behind Changes
Needed for 
#5288 
#3924
#4763

### Suggested Testing Steps
Launch DJ Hero. Note that the warning for a turntable goes away, and the game is actually playable, even without a real turntable.

seta-san commented:
Excellent. This is one of the last usb devices to get games to work. I think the other big one is the Konami ON-SAY Microphone for Japanese karaoke revolution games.

sanjay900 commented:
Only thing that isn't quite working right now is scratching, and I'm not sure how we would handle the euphoria led, for PS3 turntables its a hid output report and for Xbox turntables its based on rumble commands. 

We probably don't need to worry about PS3 turntables unless there's an easy way to get raw hid reports from them though, as they shove the crossfader and effects knob and platter buttons into accelerometer bytes which are then put in a vendor region of the hid report, so they aren't exposed via sdl.

Mrlinkwii commented:
![image](https://github.com/PCSX2/pcsx2/assets/17685135/6a1f4d74-a422-48b1-937b-afafc3cbdeed)
![image](https://github.com/PCSX2/pcsx2/assets/17685135/bacf91b3-aad3-41cf-8f5f-46a7cae24525)

works well as it can do with a ds3 controller :smile:

stenzek commented on pcsx2/USB/usb-pad/usb-pad.h:
```diff
@@ -27,6 +27,14 @@ namespace usb_pad
 		CID_STEERING_R,
 		CID_THROTTLE,
 		CID_BRAKE,
+		CID_DJ_CROSSFADER_LEFT,
```
Please try to re-use the existing enumeration values, rather than adding new values.

sanjay900 replied:
yeah i can do that, just wasn't sure what the recommendation was there

sanjay900 replied:
I didn't really see any examples of other pad types that used any other axis' besides steering wheel related stuff though.

stenzek replied:
You can reuse the sterring/throttle/etc axes for anything else.

sanjay900 replied:
But then i would be relying on the logic in GetBindValue for that id right?

sanjay900 replied:
Hows that? I just used some of the button CIDs, as there weren't going to be enough steering ones.

stenzek replied:
In that case, it may be better to create a completely separate device, rather than piggybacking off the wheels.

sanjay900 replied:
Ah yeah, good call, ill do that

stenzek replied:
Can pretty much just copy+rename usb_pad, and strip all the bits you don't need.

I know it's extra work, but it just feels a little wrong to add all these fields which are only used by one controller type. Tbh, half of the controllers in there already aren't wheels, but I didn't have the energy to duplicate them when I rewrote all the USB stuff last year.

sanjay900 replied:
Yeah, that is perfectly valid, i have done this now :). Also means i could leave a stub there for handling the euphoria led, in case someone wants to handle that on xinput using vibration commands, or if we ever get real PS3 turntable handling then it leaves it open for that too.

sanjay900 commented:
Ah, turns out that after properly centering some unused axis', scratching is now working

sanjay900 commented:
I've squashed everything down to a single commit and prefixed the commit message this time

stenzek commented on pcsx2/USB/usb-turntable/usb-turntable.cpp:
```diff
@@ -0,0 +1,488 @@
+
+namespace usb_turntable
+{
+	enum
+	{
+		STR_MANUFACTURER = 1,
+		STR_PRODUCT_MOUSE,
+		STR_PRODUCT_TABLET,
+		STR_PRODUCT_KEYBOARD,
+		STR_SERIALNUMBER,
+		STR_CONFIG_MOUSE,
+		STR_CONFIG_TABLET,
+		STR_CONFIG_KEYBOARD,
+	};
```
I can't see this enum referenced anywhere, could it be removed?

sanjay900 replied:
Oops yeah, that was copied when i cloned one of the other hid devices, removed now

stenzek reviewed (commented):
Could you put it in the `usb-pad` directory? AFAIK this is a pad-like device, unless there's going to be a bunch of others with the same characteristics.

You'll also need to update the vcxproj files. Diff is below if you're not using Windows.
```diff
diff --git a/pcsx2/pcsx2.vcxproj b/pcsx2/pcsx2.vcxproj
index 455cfd692..c8f2c9358 100644
--- a/pcsx2/pcsx2.vcxproj
+++ b/pcsx2/pcsx2.vcxproj
@@ -341,6 +341,7 @@
     <ClCompile Include="USB\usb-pad\usb-pad-sdl-ff.cpp" />
     <ClCompile Include="USB\usb-pad\usb-pad.cpp" />
     <ClCompile Include="USB\usb-pad\usb-seamic.cpp" />
+    <ClCompile Include="USB\usb-pad\usb-turntable.cpp" />
     <ClCompile Include="USB\usb-printer\usb-printer.cpp" />
     <ClCompile Include="USB\USB.cpp" />
     <ClCompile Include="VirtualMemory.cpp" />
@@ -698,6 +699,7 @@
     <ClInclude Include="USB\usb-pad\lg\lg_ff.h" />
     <ClInclude Include="USB\usb-pad\usb-pad-sdl-ff.h" />
     <ClInclude Include="USB\usb-pad\usb-pad.h" />
+    <ClInclude Include="USB\usb-pad\usb-turntable.h" />
     <ClInclude Include="USB\usb-printer\usb-printer.h" />
     <ClInclude Include="USB\USB.h" />
     <ClInclude Include="Utilities\AsciiFile.h" />
diff --git a/pcsx2/pcsx2.vcxproj.filters b/pcsx2/pcsx2.vcxproj.filters
index 06e5aacba..8b5c36a83 100644
--- a/pcsx2/pcsx2.vcxproj.filters
+++ b/pcsx2/pcsx2.vcxproj.filters
@@ -1244,6 +1244,9 @@
     <ClCompile Include="USB\usb-pad\usb-seamic.cpp">
       <Filter>System\Ps2\USB\usb-pad</Filter>
     </ClCompile>
+    <ClCompile Include="USB\usb-pad\usb-turntable.cpp">
+      <Filter>System\Ps2\USB\usb-pad</Filter>
+    </ClCompile>
     <ClCompile Include="USB\usb-pad\usb-pad-ff.cpp">
       <Filter>System\Ps2\USB\usb-pad</Filter>
     </ClCompile>
@@ -2210,6 +2213,9 @@
     <ClInclude Include="USB\usb-pad\usb-pad.h">
       <Filter>System\Ps2\USB\usb-pad</Filter>
     </ClInclude>
+    <ClInclude Include="USB\usb-pad\usb-turntable.h">
+      <Filter>System\Ps2\USB\usb-pad</Filter>
+    </ClInclude>
     <ClInclude Include="USB\usb-pad\lg\lg_ff.h">
       <Filter>System\Ps2\USB\usb-pad\lg</Filter>
     </ClInclude>
-- 
2.34.1

```

Otherwise LGTM

sanjay900 commented:
> Could you put it in the `usb-pad` directory? AFAIK this is a pad-like device, unless there's going to be a bunch of others with the same characteristics.
> 
> You'll also need to update the vcxproj files. Diff is below if you're not using Windows.
> 
> ```diff
> diff --git a/pcsx2/pcsx2.vcxproj b/pcsx2/pcsx2.vcxproj
> index 455cfd692..c8f2c9358 100644
> --- a/pcsx2/pcsx2.vcxproj
> +++ b/pcsx2/pcsx2.vcxproj
> @@ -341,6 +341,7 @@
>      <ClCompile Include="USB\usb-pad\usb-pad-sdl-ff.cpp" />
>      <ClCompile Include="USB\usb-pad\usb-pad.cpp" />
>      <ClCompile Include="USB\usb-pad\usb-seamic.cpp" />
> +    <ClCompile Include="USB\usb-pad\usb-turntable.cpp" />
>      <ClCompile Include="USB\usb-printer\usb-printer.cpp" />
>      <ClCompile Include="USB\USB.cpp" />
>      <ClCompile Include="VirtualMemory.cpp" />
> @@ -698,6 +699,7 @@
>      <ClInclude Include="USB\usb-pad\lg\lg_ff.h" />
>      <ClInclude Include="USB\usb-pad\usb-pad-sdl-ff.h" />
>      <ClInclude Include="USB\usb-pad\usb-pad.h" />
> +    <ClInclude Include="USB\usb-pad\usb-turntable.h" />
>      <ClInclude Include="USB\usb-printer\usb-printer.h" />
>      <ClInclude Include="USB\USB.h" />
>      <ClInclude Include="Utilities\AsciiFile.h" />
> diff --git a/pcsx2/pcsx2.vcxproj.filters b/pcsx2/pcsx2.vcxproj.filters
> index 06e5aacba..8b5c36a83 100644
> --- a/pcsx2/pcsx2.vcxproj.filters
> +++ b/pcsx2/pcsx2.vcxproj.filters
> @@ -1244,6 +1244,9 @@
>      <ClCompile Include="USB\usb-pad\usb-seamic.cpp">
>        <Filter>System\Ps2\USB\usb-pad</Filter>
>      </ClCompile>
> +    <ClCompile Include="USB\usb-pad\usb-turntable.cpp">
> +      <Filter>System\Ps2\USB\usb-pad</Filter>
> +    </ClCompile>
>      <ClCompile Include="USB\usb-pad\usb-pad-ff.cpp">
>        <Filter>System\Ps2\USB\usb-pad</Filter>
>      </ClCompile>
> @@ -2210,6 +2213,9 @@
>      <ClInclude Include="USB\usb-pad\usb-pad.h">
>        <Filter>System\Ps2\USB\usb-pad</Filter>
>      </ClInclude>
> +    <ClInclude Include="USB\usb-pad\usb-turntable.h">
> +      <Filter>System\Ps2\USB\usb-pad</Filter>
> +    </ClInclude>
>      <ClInclude Include="USB\usb-pad\lg\lg_ff.h">
>        <Filter>System\Ps2\USB\usb-pad\lg</Filter>
>      </ClInclude>
> -- 
> 2.34.1
> ```
> 
> Otherwise LGTM

I have moved it to usb-pad, and i have also changed the namespace to match as well, and updated the vcxproj.

stenzek commented:
Sounds good. Thanks for implementing this!
